src/basic/chase-symlinks.c

```diff
@@ -552,14 +552,22 @@ int chase_symlinks_and_opendir(
                 return 0;
         }
 
-        r = chase_symlinks(path, root, chase_flags, ret_path ? &p : NULL, &path_fd);
+        r = chase_symlinks(path, root, chase_flags, &p, &path_fd);
         if (r < 0)
                 return r;
         assert(path_fd >= 0);
 
         d = opendir(FORMAT_PROC_FD_PATH(path_fd));
```

Comment:
so, you can side-step the whole issue for directories actually, because we can convert an O_PATH fd pointing to a dir to a proper fd without `/proc/`, via `xopendirat(path_fd, ".", O_NOFOLLOW)`.

This only works for dirs, not for regular files but that's fine here.

fd_reopen() does that already as special casing.